prettyprinter-configurable/src/Text/Fixity/Internal.hs

```diff
@@ -260,19 +269,21 @@ encloseIn
     -> a
     -> a
 encloseIn parens (RenderContext dir (Fixity assocOut precOut)) (Fixity assocInn precInn) =
-    case precOut `compare` precInn of
-        LT -> id                       -- If the outer precedence is lower than the inner, then
-                                       -- do not add parens. E.g. in @Add x (Mul y z)@ the precedence
-                                       -- of @Add@ is lower than the one of @Mul@, hence there is
-                                       -- no need for parens in @x + y * z@.
-        GT -> parens                   -- If the outer precedence is greater than the inner, then
-                                       -- do add parens. E.g. in @Mul x (Add y z)@ the precedence
-                                       -- of @Mul@ is greater than the one of @Add@, hence
-                                       -- parens are needed in @x * (y + z)@.
-        EQ -> case (assocOut, dir) of  -- If precedences are equal, then judge from associativity.
-            _ | assocOut /= assocInn       -> parens  -- Associativities differ => parens are needed.
-            (LeftAssociative, ToTheLeft)   -> id      -- No need for parens in @Add (Add x y) z@
-                                                      -- which is rendered as @x + y + z@.
-            (RightAssociative, ToTheRight) -> id      -- No need for parens in @Concat xs (Concat ys zs)@
-                                                      -- which is rendered as @xs ++ ys ++ zs@.
-            _                              -> parens  -- Every other case requires parens.
+  case precOut `compare` precInn of
+    LT -> id  -- If the outer precedence is lower than the inner, then
+              -- do not add parens. E.g. in @Add x (Mul y z)@ the precedence
+              -- of @Add@ is lower than the one of @Mul@, hence there is
+              -- no need for parens in @x + y * z@.
+    GT -> parens  -- If the outer precedence is greater than the inner, then
+                  -- do add parens. E.g. in @Mul x (Add y z)@ the precedence
+                  -- of @Mul@ is greater than the one of @Add@, hence
+                  -- parens are needed in @x * (y + z)@.
+    EQ ->
+      case (assocOut, dir) of  -- If precedences are equal, then judge from associativity.
+        _ | assocOut /= assocInn       -> parens  -- Associativities differ => parens are needed.
+        (LeftAssociative, ToTheLeft)   -> id      -- No need for parens in @Add (Add x y) z@
+                                                  -- which is rendered as @x + y + z@.
+        (RightAssociative, ToTheRight) -> id      -- No need for parens in
+                                                  -- @Concat xs (Concat ys zs)@
+                                                  -- which is rendered as @xs ++ ys ++ zs@.
+        _                              -> parens  -- Every other case requires parens.
```

Comment:
Fourmolu disabled. 
The trade-off here: current formatting isn't diff-friendly (comment blocks shift horizontally depending on the lhs expression length) but is fitted to this particular code width thus making it more compact.

My preference is in making it less fragile by placing comments above the expressions with the same indentation level, such that it doesn't depend on the expression length:

```suggestion
  case precOut `compare` precInn of
    LT -> 
      -- If the outer precedence is lower than the inner, then
      -- do not add parens. E.g. in @Add x (Mul y z)@ the precedence
      -- of @Add@ is lower than the one of @Mul@, hence there is
      -- no need for parens in @x + y * z@.
      id
    GT ->
      -- If the outer precedence is greater than the inner, then
      -- do add parens. E.g. in @Mul x (Add y z)@ the precedence
      -- of @Mul@ is greater than the one of @Add@, hence
      -- parens are needed in @x * (y + z)@.
      parens
    EQ ->
      -- If precedences are equal, then judge from associativity.
      case (assocOut, dir) of 
        _ | assocOut /= assocInn ->
          -- Associativities differ => parens are needed.
          parens 
        (LeftAssociative, ToTheLeft) ->
          -- No need for parens in @Add (Add x y) z@
          -- which is rendered as @x + y + z@.
          id 
        (RightAssociative, ToTheRight) -> 
          -- No need for parens in
          -- @Concat xs (Concat ys zs)@
          -- which is rendered as @xs ++ ys ++ zs@.
          id
        _ ->
          -- Every other case requires parens.
          parens 
```

Such code trades vertical space (which I have plenty of) for precious horizontal space (which I like to save).